File: Source/Extensions/Tessa.Extensions.Default.Client/Workflow/KrProcess/CommandInterpreter/AdvancedDialogCommandHandler.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Input;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess.ClientCommandInterpreter;
using Tessa.Extensions.Platform.Client.UI;
using Tessa.Files;
using Tessa.Platform;
using Tessa.Platform.Runtime;
using Tessa.Platform.Validation;
using Tessa.UI;
using Tessa.UI.Cards;

namespace Tessa.Extensions.Default.Client.Workflow.KrProcess.CommandInterpreter
{
    /// <summary>
    /// Базовый класс обработчика клиентской команды отображения диалога.
    /// </summary>
    public abstract class AdvancedDialogCommandHandler : ClientCommandHandlerBase
    {
        #region Fields

        private readonly Func<IAdvancedCardDialogManager> createAdvancedCardDialogManagerFunc;

        private readonly ISession session;

        private readonly ICardMetadata cardMetadata;
        
        #endregion

        #region Constructor

        protected AdvancedDialogCommandHandler(
            Func<IAdvancedCardDialogManager> createAdvancedCardDialogManagerFunc,
            ISession session,
            ICardMetadata cardMetadata)
        {
            this.createAdvancedCardDialogManagerFunc = NotNullOrThrow(createAdvancedCardDialogManagerFunc);
            this.session = NotNullOrThrow(session);
            this.cardMetadata = NotNullOrThrow(cardMetadata);
        }

        #endregion

        #region Base Overrides

        /// <inheritdoc />
        public override Task Handle(
            IClientCommandHandlerContext context)
        {
            var coSettings = this.PrepareDialogCommand(context);

            if (coSettings is null)
            {
                return Task.CompletedTask;
            }

            var editor = UIContext.Current.CardEditor;
            if (editor is not null)
            {
                editor.Info[CardTaskDialogUIExtension.DialogNonTaskCompletionOptionSettingsKey] = coSettings;
                editor.Info[CardTaskDialogUIExtension.OnDialogButtonPressedKey] =
                    (Func<ICardEditorModel, CardTaskCompletionOptionSettings, string?, bool, ValueTask>) (
                        async (
                            dialogCardEditor,
                            cos,
                            buttonName,
                            completeTask) => await this.CompleteDialogAsync(
                                dialogCardEditor,
                                editor,
                                cos,
                                buttonName,
                                completeTask,
                                context));
            }
            else
            {
                _ = this.ShowGlobalDialogAsync(
                    coSettings,
                    context);
            }

            return Task.CompletedTask;
        }

        #endregion

        #region Protected Abstract Methods

        /// <summary>
        /// Метод для подготовки диалога для выполнения.
        /// </summary>
        /// <param name="context">Контекст обработки клиентской команды.</param>
        /// <returns>Информация для формирования диалога, или значение <see langword="null"/>, если невозможно сформировать диалог.</returns>
        protected abstract CardTaskCompletionOptionSettings? PrepareDialogCommand(IClientCommandHandlerContext context);

        /// <summary>
        /// Метод выполнения диалога.
        /// </summary>
        /// <param name="actionResult">Результат выполнения диалога.</param>
        /// <param name="context">Контекст команды диалога.</param>
        /// <param name="cardEditor">Редактор карточки диалога.</param>
        /// <param name="parentCardEditor">Редактор карточки, для которой открывается диалог, если диалог открывается в рамках карточки.</param>
        /// <returns>Значение <see langword="true"/>, если необходимо закрыть диалог, иначе - <see langword="false"/>.</returns>
        protected abstract ValueTask<bool> CompleteDialogCoreAsync(
            CardTaskDialogActionResult actionResult,
            IClientCommandHandlerContext context,
            ICardEditorModel cardEditor,
            ICardEditorModel? parentCardEditor = null);

        #endregion

        #region Private Methods

        /// <summary>
        /// Отображает карточку в окне диалога.
        /// </summary>
        /// <param name="coSettings">Параметры диалога.</param>
        /// <param name="context">Контекст обработки клиентской команды.</param>
        /// <returns>Асинхронная задача.</returns>
        private async Task ShowGlobalDialogAsync(
            CardTaskCompletionOptionSettings coSettings,
            IClientCommandHandlerContext context)
        {
            var info = coSettings.Info;
            if (coSettings.PreparedNewCard is not null
                && coSettings.PreparedNewCardSignature is not null)
            {
                info[CardHelper.NewCardBilletKey] = coSettings.PreparedNewCard;
                info[CardHelper.NewCardBilletSignatureKey] = coSettings.PreparedNewCardSignature;
            }

            info[CardTaskDialogHelper.StoreMode] = Int32Boxes.Box((int) coSettings.StoreMode);

            await this.CreateNewCardAsync(
                this.createAdvancedCardDialogManagerFunc(),
                coSettings,
                info,
                async ctx =>
                {
                    ctx.Editor.Context.SetDialogClosingAction((uiCtx, eventArgs) =>
                    {
                        if (uiCtx.CardEditor.OperationInProgress)
                        {
                            eventArgs.Cancel = true;
                        }
                        else if (!uiCtx.CardEditor.IsClosed
                                && (!coSettings.IsCloseWithoutConfirmation
                                || uiCtx.CardEditor.CardModel.Card.HasChanges()))
                        {
                            var dialogResult = Keyboard.Modifiers.HasNot(ModifierKeys.Alt)
                                && TessaDialog.Confirm("$KrProcess_Dialog_ConfirmClose");

                            if (!dialogResult)
                            {
                                eventArgs.Cancel = true;
                            }
                        }

                        return TaskBoxes.False;
                    });
                },
                ctx =>
                {
                    // Подготовка диалога к использованию.
                    this.PrepareDialog(ctx.Editor, coSettings, context);

                    return ValueTask.CompletedTask;
                },
                CancellationToken.None).ConfigureAwait(false);
        }

        private void PrepareDialog(
           ICardEditorModel editor,
           CardTaskCompletionOptionSettings coSettings,
           IClientCommandHandlerContext context)
        {
            editor.StatusBarIsVisible = false;

            var toolbarActions = editor.Toolbar.Actions;
            var bottomActions = editor.BottomToolbar.Actions;
            var bottomButtons = editor.BottomDialogButtons;

            toolbarActions.Clear();
            bottomActions.Clear();
            bottomButtons.Clear();

            foreach (var actionInfo in coSettings.Buttons)
            {
                var name = actionInfo.Name;
                var icon = string.IsNullOrEmpty(actionInfo.Icon)
                    ? null
                    : editor.Toolbar.CreateIcon(actionInfo.Icon);

                bool CanExecuteButtonAction(object? _)
                {
                    return !editor.BlockingOperationInProgress;
                }

                switch (actionInfo.CardButtonType)
                {
                    case CardButtonType.ToolbarButton:
                        var topActionAsync = this.GetButtonActionAsync(editor, coSettings, context, actionInfo);
                        toolbarActions.Add(new CardToolbarAction(
                            name,
                            actionInfo.Caption,
                            icon,
                            // ReSharper disable once AccessToModifiedClosure
                            new DelegateCommand(async _ => await topActionAsync(null), CanExecuteButtonAction),
                            order: actionInfo.Order
                        ));
                        break;
                    case CardButtonType.BottomToolbarButton:
                        var bottomActionAsync = this.GetButtonActionAsync(editor, coSettings, context, actionInfo);
                        bottomActions.Add(new CardToolbarAction(
                            name,
                            actionInfo.Caption,
                            icon,
                            // ReSharper disable once AccessToModifiedClosure
                            new DelegateCommand(async _ => await bottomActionAsync(null), CanExecuteButtonAction),
                            order: actionInfo.Order
                        ));
                        break;
                    case CardButtonType.BottomDialogButton:
                        bottomButtons.Add(new UIButton(
                            actionInfo.Caption,
                            this.GetButtonActionAsync(editor, coSettings, context, actionInfo),
                            () => CanExecuteButtonAction(null)
                        ));
                        break;
                    default:
                        throw new ArgumentOutOfRangeException(
                            $"{nameof(actionInfo)}.{nameof(actionInfo.CardButtonType)}",
                            actionInfo.CardButtonType,
                            null);
                }
            }
        }

        private Func<UIButton?, ValueTask> GetButtonActionAsync(
            ICardEditorModel editor,
            CardTaskCompletionOptionSettings coSettings,
            IClientCommandHandlerContext context,
            CardTaskDialogButtonInfo buttonInfo)
        {
            return buttonInfo.Cancel
                ? async _ => await editor.CloseAsync()
                : async _ => await this.CompleteDialogAsync(
                    editor,
                    null,
                    coSettings,
                    buttonInfo.Name,
                    buttonInfo.CompleteDialog,
                    context);
        }

        private async Task CompleteDialogAsync(
            ICardEditorModel dialogCardEditor,
            ICardEditorModel? parentCardEditor,
            CardTaskCompletionOptionSettings coSettings,
            string? buttonName,
            bool completeDialog,
            IClientCommandHandlerContext context)
        {
            var closeDialog = false;
            using (dialogCardEditor.SetOperationInProgress(true))
            {
                var dialogCard = dialogCardEditor.CardModel.Card.Clone();
                var files = dialogCardEditor.CardModel.FileContainer.Files;

                var validationResult = await this.PrepareFilesForStoreAsync(
                    dialogCard,
                    files);

                await TessaDialog.ShowNotEmptyAsync(validationResult);
                if (validationResult.HasErrors)
                {
                    return;
                }

                var actionResult = new CardTaskDialogActionResult
                {
                    MainCardID = Guid.Empty,
                    PressedButtonName = buttonName,
                    StoreMode = coSettings.StoreMode,
                    KeepFiles = coSettings.KeepFiles,
                    CompleteDialog = completeDialog,
                };
                actionResult.SetDialogCard(dialogCard);

                closeDialog = await this.CompleteDialogCoreAsync(
                    actionResult,
                    context,
                    dialogCardEditor,
                    parentCardEditor);
            }
            if (closeDialog)
            {
                await dialogCardEditor.CloseAsync();
            }
        }

        /// <summary>
        /// Асинхронно задаёт контент указанных файлов в соответствующие <see cref="CardFile.Info"/> карточки файлов.
        /// </summary>
        /// <param name="dialogCard">Карточка диалога.</param>
        /// <param name="files">Коллекция файлов.</param>
        /// <param name="cancellationToken">Объект, посредством которого можно отменить асинхронную задачу.</param>
        /// <returns>Результат выполнения операции, агрегированный для всех файлов.</returns>
        private async Task<ValidationResult> PrepareFilesForStoreAsync(
            Card dialogCard,
            ICollection<IFile> files,
            CancellationToken cancellationToken = default)
        {
            var validationResults = new ValidationResultBuilder
            {
                await files.EnsureAllContentModifiedAsync(cancellationToken)
            };

            if (validationResults.IsSuccessful())
            {
                await CardTaskDialogHelper.SetFileContentToInfoAsync(
                    dialogCard,
                    files,
                    this.session,
                    validationResults,
                    cancellationToken);
            }

            return validationResults.Build();
        }

        /// <summary>
        /// Асинхронно создаёт и открывает карточку в диалоге. Карточка создаётся в режиме по умолчанию или по шаблону.
        /// </summary>
        /// <param name="advancedCardDialogManager"><inheritdoc cref="IAdvancedCardDialogManager" path="/summary"/></param>
        /// <param name="completionOptionSettings"><inheritdoc cref="CardTaskCompletionOptionSettings" path="/summary"/></param>
        /// <param name="info"><inheritdoc cref="CreateCardOptions.Info" path="/summary"/></param>
        /// <param name="cardEditorModifierActionAsync"><inheritdoc cref="CreateCardOptions.CardEditorModifierActionAsync" path="/summary"/></param>
        /// <param name="cardModifierActionAsync"><inheritdoc cref="CreateCardOptions.CardModifierActionAsync" path="/summary"/></param>
        /// <param name="cancellationToken">Объект, посредством которого можно отменить асинхронную задачу.</param>
        /// <returns>Асинхронная задача.</returns>
        private async Task CreateNewCardAsync(
            IAdvancedCardDialogManager advancedCardDialogManager,
            CardTaskCompletionOptionSettings completionOptionSettings,
            Dictionary<string, object?>? info = null,
            CardEditorCreationActionAsync? cardEditorModifierActionAsync = null,
            CardEditorCreationActionAsync? cardModifierActionAsync = null,
            CancellationToken cancellationToken = default)
        {
            var dialogDisplayValue = completionOptionSettings.DisplayValue;
            if (completionOptionSettings.CardNewMethod == CardTaskDialogNewMethod.Default &&
                string.IsNullOrEmpty(dialogDisplayValue) &&
                (await this.cardMetadata.GetCardTypesAsync(cancellationToken)).TryGetValue(completionOptionSettings.DialogTypeID, out var dialogType))
            {
                dialogDisplayValue = dialogType.Caption;
            }
            
            var createCardOptions = new CreateCardOptions
            {
                Info = info,
                DisplayValue = dialogDisplayValue,
                UIContext = UIContext.Current,
                CardEditorModifierActionAsync = cardEditorModifierActionAsync,
                CardModifierActionAsync = async ctx =>
                {
                    ctx.Editor.DialogName = completionOptionSettings.DialogName;

                    if (cardModifierActionAsync is not null)
                    {
                        await cardModifierActionAsync(ctx);
                    }
                },
                SaveCreationRequest = false,
            };

            switch (completionOptionSettings.CardNewMethod)
            {
                case CardTaskDialogNewMethod.Default:
                    await advancedCardDialogManager.CreateCardAsync(
                        completionOptionSettings.DialogTypeID,
                        options: createCardOptions,
                        cancellationToken: cancellationToken).ConfigureAwait(false);
                    break;
                case CardTaskDialogNewMethod.Template:
                    await advancedCardDialogManager.CreateFromTemplateAsync(
                        completionOptionSettings.DialogTypeID,
                        createCardOptions,
                        cancellationToken).ConfigureAwait(false);
                    break;
                default:
                    throw new ArgumentOutOfRangeException(
                        nameof(CardTaskCompletionOptionSettings.CardNewMethod),
                        completionOptionSettings.CardNewMethod,
                        default);
            }
        }

        #endregion

    }
}

File: WebClient SDK/src/default/workflow/krProcess/commandInterpreter/advancedDialogCommandHandler.ts
import { ClientCommandHandlerBase } from 'tessa/workflow/krProcess/clientCommandInterpreter';
import { IClientCommandHandlerContext } from 'tessa/workflow/krProcess';
import {
  CardButtonType,
  CardTaskCompletionOptionSettings,
  CardTaskDialogActionResult,
  systemKeyPrefix,
  Card,
  CardTaskDialogNewMethod,
  CardTaskDialogButtonInfo
} from 'tessa/cards';
import {
  IUIContext,
  showConfirm,
  showNotEmpty,
  UIButton,
  UIContext,
  createCardEditorModel,
  tryGetFromInfo
} from 'tessa/ui';
import {
  AdvancedCardDialogManager,
  CardEditorCreationContext,
  CardToolbarAction,
  ICardEditorModel
} from 'tessa/ui/cards';
import { createTypedField, DotNetType, Guid } from 'tessa/platform';
import { IStorage } from 'tessa/platform/storage';
import { IFile, setFileContentToInfo } from 'tessa/files';
import { CardRequestExtensionContext } from 'tessa/cards/extensions';
import { getTessaIcon } from 'common/utility';
import { ValidationResult, ValidationResultBuilder } from 'tessa/platform/validation';
import { ArgumentOutOfRangeError } from 'tessa/platform/errors';
import { CreateCardArg } from 'tessa/ui/uiHost/common';
import { MetadataStorage } from 'tessa/metadataStorage';

/**
 * Базовый класс обработчика клиентской команды отображения диалога.
 */
export abstract class AdvancedDialogCommandHandler extends ClientCommandHandlerBase {
  //#region methods

  async handle(context: IClientCommandHandlerContext): Promise<void> {
    const coSettings = this.prepareDialogCommand(context);
    if (!coSettings) {
      return;
    }

    // в ТК при вызове диалога через тайл или при нажатии на кнопку в тулбаре UIContext не сохраняется.
    // в ЛК UIContext, который создается в showGlobalDialogAsync, сохраняется и доступен в расширениях,
    // поэтому поведение в этой части расширения может отличаться.
    // Для того чтобы обойти эту проблему в WeAdvancedDialogCommandHandler и KrAdvancedDialogCommandHandler прокидывается флажок
    // Когда этот флажок есть в CardRequestExtensionContext (и только для этого типа реквестов), то мы пропускаем проверку UIContext.
    let skipEditor = false;
    if (context.outerContext && context.outerContext instanceof CardRequestExtensionContext) {
      skipEditor = tryGetFromInfo(
        context.outerContext.request.info,
        systemKeyPrefix + 'WebAdvancedDialogCommandSkipUIContextFlag',
        false
      );
    }

    const editor = UIContext.current.cardEditor;
    if (editor && !skipEditor) {
      editor.info[systemKeyPrefix + 'CardEditorCompletionOptionSettings'] = coSettings;
      editor.info[systemKeyPrefix + 'CardEditorCompletionOptionSettingsOnButtonPressed'] = async (
        dialogCardEditor: ICardEditorModel,
        cos: CardTaskCompletionOptionSettings,
        buttonName: string | null,
        completeTask: boolean
      ) =>
        await this.completeDialog(dialogCardEditor, editor, cos, buttonName, completeTask, context);
    } else {
      setTimeout(() => this.showGlobalDialog(coSettings, context));
    }
  }

  //#endregion

  //#region protected methods

  /**
   * Метод для подготовки диалога для выполнения.
   *
   * @param {IClientCommandHandlerContext} context Контекст обработки клиентской команды.
   * @returns {CardTaskCompletionOptionSettings | null} Информация для формирования диалога, или значение null, если невозможно сформировать диалог.
   */
  protected abstract prepareDialogCommand(
    context: IClientCommandHandlerContext
  ): CardTaskCompletionOptionSettings | null;

  /**
   * Метод выполнения диалога.
   *
   * @param {CardTaskDialogActionResult} actionResult Результат выполнения диалога.
   * @param {IClientCommandHandlerContext} context Контекст команды диалога.
   * @param {ICardEditorModel} cardEditor Редактор карточки диалога.
   * @param {(ICardEditorModel | null)} parentCardEditor Редактор карточки, для которой открывается диалог, если диалог открывается в рамках карточки.
   * @returns {Promise<boolean>} Значение true, если необходимо закрыть диалог, иначе false.
   */
  protected abstract completeDialogCore(
    actionResult: CardTaskDialogActionResult,
    context: IClientCommandHandlerContext,
    cardEditor: ICardEditorModel,
    parentCardEditor: ICardEditorModel | null
  ): Promise<boolean>;

  //#endregion

  //#region private methods

  /**
   * Отображает карточку в окне диалога.
   *
   * @param {CardTaskCompletionOptionSettings} coSettings Параметры диалога.
   * @param {IClientCommandHandlerContext} context Контекст обработки клиентской команды.
   */
  private async showGlobalDialog(
    coSettings: CardTaskCompletionOptionSettings,
    context: IClientCommandHandlerContext
  ): Promise<void> {
    const info = coSettings.info;
    if (coSettings.preparedNewCard && coSettings.preparedNewCardSignature) {
      info[systemKeyPrefix + 'NewBilletCard'] = createTypedField(
        coSettings.preparedNewCard,
        DotNetType.Binary
      );
      info[systemKeyPrefix + 'NewBilletCardSignature'] = createTypedField(
        coSettings.preparedNewCardSignature,
        DotNetType.Binary
      );
    }
    info[systemKeyPrefix + 'StoreMode'] = createTypedField(coSettings.storeMode, DotNetType.Int32);

    const contextInstance = UIContext.create(
      new UIContext({
        cardEditor: createCardEditorModel(),
        actionOverridings: AdvancedCardDialogManager.instance.createUIContextActionOverridings()
      })
    );
    try {
      await AdvancedDialogCommandHandler.createNewCard(
        coSettings,
        info,
        ctx => {
          ctx.editor.context.info[systemKeyPrefix + 'DialogClosingAction'] = async (
            uiCtx: IUIContext,
            args: { cancel: boolean }
          ) => {
            if (uiCtx.cardEditor!.operationInProgress) {
              args.cancel = true;
            } else if (
              !uiCtx.cardEditor!.isClosed &&
              (!coSettings.isCloseWithoutConfirmation ||
                uiCtx.cardEditor!.cardModel!.card.hasChanges())
            ) {
              // TODO: Keyboard.Modifiers.HasNot(ModifierKeys.Alt)
              const dialogResult = await showConfirm('$KrProcess_Dialog_ConfirmClose');
              if (!dialogResult) {
                args.cancel = true;
              }
            }
            return false;
          };
        },
        ctx => {
          // Подготовка диалога к использованию.
          this.prepareDialog(ctx.editor, coSettings, context);
        }
      );
    } finally {
      contextInstance.dispose();
    }
  }

  private prepareDialog(
    editor: ICardEditorModel,
    coSettings: CardTaskCompletionOptionSettings,
    context: IClientCommandHandlerContext
  ) {
    editor.statusBarIsVisible = false;

    editor.toolbar.clearItems();
    editor.bottomToolbar.clearItems();
    editor.bottomDialogButtons.length = 0;

    for (const actionInfo of coSettings.buttons) {
      const name = actionInfo.name;
      const icon = !!actionInfo.icon ? getTessaIcon(actionInfo.icon) : '';

      switch (actionInfo.cardButtonType) {
        case CardButtonType.BottomToolbarButton:
          editor.bottomToolbar.addItem(
            new CardToolbarAction({
              name,
              caption: actionInfo.caption,
              icon: icon,
              command: this.getButtonAction(editor, coSettings, context, actionInfo),
              order: actionInfo.order
            })
          );
          break;
        case CardButtonType.ToolbarButton:
          editor.toolbar.addItem(
            new CardToolbarAction({
              name,
              caption: actionInfo.caption,
              icon: icon,
              command: this.getButtonAction(editor, coSettings, context, actionInfo),
              order: actionInfo.order
            })
          );
          break;
        case CardButtonType.BottomDialogButton:
          editor.bottomDialogButtons.push(
            new UIButton(
              actionInfo.caption,
              this.getButtonAction(editor, coSettings, context, actionInfo)
            )
          );
          break;
        default:
          throw new ArgumentOutOfRangeError('actionInfo.cardButtonType', actionInfo.cardButtonType);
      }
    }
  }

  private getButtonAction(
    editor: ICardEditorModel,
    coSettings: CardTaskCompletionOptionSettings,
    context: IClientCommandHandlerContext,
    buttonInfo: CardTaskDialogButtonInfo
  ): () => Promise<void> {
    return buttonInfo.cancel
      ? async () => {
          await editor.close();
        }
      : async () =>
          await this.completeDialog(
            editor,
            null,
            coSettings,
            buttonInfo.name,
            buttonInfo.completeDialog,
            context
          );
  }

  private async completeDialog(
    dialogCardEditor: ICardEditorModel,
    parentCardEditor: ICardEditorModel | null,
    coSettings: CardTaskCompletionOptionSettings,
    buttonName: string | null,
    completeDialog: boolean,
    context: IClientCommandHandlerContext
  ) {
    let closeDialog = false;

    await dialogCardEditor.setOperationInProgress(async () => {
      const dialogCard = dialogCardEditor.cardModel!.card.clone();

      const validationResult = await AdvancedDialogCommandHandler.prepareFilesForStore(
        dialogCard,
        dialogCardEditor.cardModel!.fileContainer.files
      );

      await showNotEmpty(validationResult);
      if (validationResult.hasErrors) {
        return;
      }

      const actionResult = new CardTaskDialogActionResult();
      actionResult.mainCardId = Guid.empty;
      actionResult.pressedButtonName = buttonName;
      actionResult.storeMode = coSettings.storeMode;
      actionResult.keepFiles = coSettings.keepFiles;
      actionResult.completeDialog = completeDialog;
      actionResult.setDialogCard(dialogCard);

      closeDialog = await this.completeDialogCore(
        actionResult,
        context,
        dialogCardEditor,
        parentCardEditor
      );
    });

    if (closeDialog) {
      await dialogCardEditor.close();
    }
  }

  /**
   * Задаёт контент указанных файлов в соответствующие CardFile.Info карточки файлов.
   *
   * @param {Card} dialogCard Карточка диалога.
   * @param {Readonly<IFile[]>} files Коллекция файлов.
   * @returns {Promise<ValidationResult>} Результат выполнения операции, агрегированный для всех файлов.
   */
  private static async prepareFilesForStore(
    dialogCard: Card,
    files: Readonly<IFile[]>
  ): Promise<ValidationResult> {
    const validationResult = new ValidationResultBuilder();
    for (const file of files) {
      validationResult.add(await file.ensureContentModified());
    }

    if (validationResult.isSuccessful) {
      await setFileContentToInfo(dialogCard, files, validationResult);
    }
    return validationResult.build();
  }

  /**
   * Создаёт и открывает карточку в диалоге. Карточка создаётся в режиме по умолчанию или по шаблону.
   *
   * @param {CardTaskCompletionOptionSettings} dialogSettings {@link CardTaskCompletionOptionSettings}
   * @param {IStorage} info {@link IStorage}
   * @param {(context: CardEditorCreationContext) => void} cardEditorModifierAction
   * @param {(context: CardEditorCreationContext) => void} cardModifierAction
   */
  private static async createNewCard(
    dialogSettings: CardTaskCompletionOptionSettings,
    info: IStorage,
    cardEditorModifierAction: (context: CardEditorCreationContext) => void,
    cardModifierAction: (context: CardEditorCreationContext) => void
  ): Promise<void> {
    const cardNewMethod = dialogSettings.cardNewMethod;

    let dialogDisplayValue = dialogSettings.displayValue;
    if (
      cardNewMethod === CardTaskDialogNewMethod.Default &&
      (!dialogDisplayValue || dialogDisplayValue === '')
    ) {
      const dialogType = MetadataStorage.instance.cardMetadata.getCardTypeById(
        dialogSettings.dialogTypeId
      );
      if (dialogType) {
        dialogDisplayValue = dialogType.caption ?? undefined;
      }
    }

    const args: CreateCardArg = {
      cardTypeId: dialogSettings.dialogTypeId,
      info: info,
      displayValue: dialogDisplayValue,
      context: UIContext.current,
      cardEditorModifierAction: cardEditorModifierAction,
      cardModifierAction: (ctx: CardEditorCreationContext) => {
        ctx.editor.dialogName = dialogSettings.dialogName;
        cardModifierAction(ctx);
      },
      dialogOptions: {
        dialogName: dialogSettings.dialogName
      },
      saveCreationRequest: false
    };

    switch (cardNewMethod) {
      case CardTaskDialogNewMethod.Default:
        await AdvancedCardDialogManager.instance.createCard(args);
        break;

      case CardTaskDialogNewMethod.Template:
        await AdvancedCardDialogManager.instance.createFromTemplate(
          dialogSettings.dialogTypeId,
          args
        );
        break;

      default:
        throw new ArgumentOutOfRangeError(
          'CardTaskCompletionOptionSettings.cardNewMethod',
          cardNewMethod
        );
    }
  }

  //#endregion
}
